### PR TITLE
Add various git completions

### PR DESCRIPTION
**Pull request by cstyles**
## Description

Adds completions for the following git commands:
- `range-diff`
- `stash push`
- `reflog`
- `rev-parse`

And I threw in some miscellaneous completions for `git log --date=human` and `git reset`.

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: share/completions/git.fish [hunk 1]
```diff
@@ -1577,13 +1598,18 @@ complete -f -c git -n '__fish_git_using_command stash; and __fish_git_stash_not_
...
 complete -f -c git -n '__fish_git_using_command stash; and __fish_git_stash_not_using_subcommand' -a branch -d 'Create a new branch from a stash'
+complete -f -c git -n '__fish_git_using_command stash; and __fish_git_stash_not_using_subcommand' -a push -d 'Create a new stash with given files'
 
 complete -f -c git -n '__fish_git_stash_using_command apply' -a '(__fish_git_complete_stashes)'
 complete -f -c git -n '__fish_git_stash_using_command branch' -a '(__fish_git_complete_stashes)'
 complete -f -c git -n '__fish_git_stash_using_command drop' -a '(__fish_git_complete_stashes)'
 complete -f -c git -n '__fish_git_stash_using_command pop' -a '(__fish_git_complete_stashes)'
 complete -f -c git -n '__fish_git_stash_using_command show' -a '(__fish_git_complete_stashes)'
 
+complete -f -c git -n '__fish_git_stash_using_command push' -a '(__fish_git_files modified)'
```
Comment (krobelus) [inline share/completions/git.fish]:
```suggestion
complete -f -c git -n '__fish_git_stash_using_command push' -a '(__fish_git_files modified deleted)'
```

File: share/completions/git.fish [hunk 2]
```diff
@@ -1435,10 +1445,15 @@ complete -f -c git -n '__fish_git_using_command rebase' -l no-ff -d 'No fast-for
 # This actually takes script for $SHELL, but completing that is... complicated.
 complete -r -c git -n '__fish_git_using_command rebase' -l exec -d 'Execute shellscript'
 
+### reflog
```
Comment (krobelus) [inline share/completions/git.fish]: Maybe we want to eventually support reflog's subcommands (other than the default `show`), but `show` probably accounts for a large majority of uses so I think this is fine.

File: share/completions/git.fish [hunk 3]
```diff
@@ -1435,10 +1445,15 @@ complete -f -c git -n '__fish_git_using_command rebase' -l no-ff -d 'No fast-for
...
+complete -f -c git -n '__fish_git_needs_command' -a reflog -d 'Manage reflog information'
+complete -f -c git -n '__fish_git_using_command reflog' -a '(__fish_git_branches)'
+complete -f -c git -n '__fish_git_using_command reflog' -a '(__fish_git_heads)' -d 'Head'
+
 ### reset
 complete -c git -n '__fish_git_needs_command' -a reset -d 'Reset current HEAD to the specified state'
 complete -f -c git -n '__fish_git_using_command reset' -l hard -d 'Reset files in working directory'
-complete -c git -n '__fish_git_using_command reset' -a '(__fish_git_branches)'
+complete -c git -n '__fish_git_using_command reset; and not contains -- -- (commandline -op)' -a '(__fish_git_branches)'
```
Comment (krobelus) [inline share/completions/git.fish]: I think we can use `commandline -c` (`--cut-at-cursor`) to keep completing refs if the cursor is left of the `--`.
You can also add the flag `-c` to all other occurrences of `not contains -- -- (commandline -op)` in this file.

Comment (cstyles) [reply share/completions/git.fish]: Neat! Thanks for the tip.

File: share/completions/git.fish (hunk 2 above)
Comment (cstyles) [reply share/completions/git.fish]: Good idea. I added completions for the subcommands in 1f28825c276224a8752aacef1a548c2a32db65c4

**Comment by cstyles** [comment]
I addressed the feedback and pushed the changes. While fixing things up (and using git to do so), I discovered some missing completions for `git-commit` so I went ahead and pushed them to this PR (793f56a9ef05d5776af106a79749ec8ec0b5b865). I'd appreciate it someone could take a look at those too.

**Comment by krobelus** [comment]
Thanks for the great work! Merged as the commits below.
I just amended 12a5dd219 to use `-x` instead of `-x` (though it doesn't really change the behavior)

ffb551fc0 master Add --cut-at-cursor option to `commandline -op` calls in git completions
12a5dd219 Add completions for git-commit
cc84dc751 Add completions for git rev-parse
486bc71cf Add completions for git reflog
9384801e3 Add completions for `push` subcommand to git stash
63e840995 git-reset: Don't suggest branch completions if -- is present
3de3a34e7 Add completion for `git log --date=human`
bd7130878 Add git completions for range-diff

**Pull request closed without merging**